homeassistant/components/mastodon/config_flow.py

```diff
@@ -127,6 +143,51 @@ async def async_step_user(
 
         return self.show_user_form(user_input, errors)
 
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.base_url = entry_data[CONF_BASE_URL]
+        self.config_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm reauth dialog."""
+        errors: dict[str, str] = {}
+        if user_input:
+            _, _, errors = await self.hass.async_add_executor_job(
+                self.check_connection,
+                self.base_url,
+                user_input[CONF_CLIENT_ID],
+                user_input[CONF_CLIENT_SECRET],
+                user_input[CONF_ACCESS_TOKEN],
+            )
+
+            if not errors:
+                assert self.config_entry
+                # Access token only used to auth but as client_id originally captured and
+                # used as unique_id we capture again/validate to avoid switching accounts.
+                if self.config_entry.unique_id == user_input[CONF_CLIENT_ID]:
```

Comment:
This one is awkward, the access token is actually all that is needed to authenticate, but as we are using the client id as our unique id there's potential to create a new client id/access token and apply it to the this unique id.
It's messy and once the yaml import is removed in 6 months I think I can revisit the auth mechanism.

Reply:
In that case I would just require the access token. If people want to use new client_id and secret, I think a reconfigure flow is more appropriate

Reply:
I tried that, but you cannot get the client_id from the library to validate we're using the same one for the access token.
Without this we could have two accounts on the same mastodon instance, and swap the access tokens without any way of validating.
At the moment this is the only way to ensure the account is the same.

Reply:
No, because the api client should only take an access token OR a client details (latter for oAuth), it's a bit misleading the way it was implemented.
There is the potential to change the access token to one for a different account and as the library doesn't validate the client details it would still pass, and then you would have a unique id/client id pointing to a different account.
I wanted to provide a way to re-auth but given the current state it may just be better to ignore for now, it's a work around and not a nice one.

I was looking at the oAuth for Nest which has unique URL's, it would perhaps be better to wait till the config.yaml migration is removed next year and move to oAuth and get users to re-authenticate with oAuth.

Reply:
So the thing what `nest` has, which `mastodon` doesn't have, is a single oauth endpoint. With Mastodon you can setup your own server and then you have to do oAuth with that server instead of a current one.

But to get back at this, we don't use the client_id and secret to validate the access token?

Reply:
nest does have a dynamic url for a project ID but yes I'll park that one for now as not looked in detail.

No, the access token is a long lived token, you only need the server url and the token and it can authenticate the account.

Reply:
Sorry this has turned into a discussion more than a PR.

One way to fix this would be to migrate the config entry to lookup/include the account name in data, then rather than validate unique_id we could validate the account name/instance returned as in the above example comment.

Reply:
I am now looking at nest and this was actually the 1% option possibility that I had in mind how it could work but didn't check yet. Honestly, should we migrate to oAuth? It probably will be easier in 6 months, but for now I don't think its impossible

Reply:
The more I'm thinking about this the migration to oAuth would be better, learning how the token works I'm a little worried that this may be removed from Mastodon long term for security.
The token doesn't change unless users delete the app registration so it's hopefully uncommon that a reauth is needed.
It will take me a while to pick apart the relevant pieces of nest but it's a good starting point if a little more complex than we need here.
I will close this PR, thanks for the discussion I had doubts on the best way to approach this but until I had some code down it was hard to explain.